### PR TITLE
Add support for async plugin `build.onResolve` at runtime

### PR DESCRIPTION
**Pull request by gungun974**
### What does this PR do?

This PR addresses my issue outlined in #5314.

The problem with #5314 is that the Bun runtime does not allow plugins to handle asynchronous imports.
I personally need this feature, so I decided to create this PR to implement it for all the Bun users.
I also planed after this PR to create a Bun plugin capable of importing image and do some image operation before building and live.

Although I'm not really familiar with Zig, I have knowledge in C/C++ and Rust and I know how computer memory can been slow some time. So I try the best to not create any useless variable.
Please Verify I didn't anything is not idiomatic to Zig and will hurt de performance of Bun.

- [x] Code changes

### How did you verify your code works?

- [x] I have written some tests for the bundler for verify if it's still works.
- [x] I have written some tests for the runtime plugin to verify if I didn't broke the none async `builder.onResolve` and if the async `build.onResolve` work as expected
- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be ***(But since I'm not a zig expert so please verify)***
- [x] I or my editor ran `zig fmt` on the changed files
- [x] I included a test for the new code, or an existing test covers it
- [x] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed

### CODE REVIEW

**Comment by gungun974** [comment]
I may have managed to add unit tests in the end but I didn't understand how I could run them on my machine.

When I use Bun test, I notice that my tests are done with my installed version of bun only.

In short I may be the worst person to made a PR on this repository but hey, you know it's the effort that counts and the willingness ^^

**Comment by gungun974** [comment]
Like said, I made this plugin, the [Bun Image Transform](https://github.com/gungun974/bun-image-transform) to easily convert and transform image on import.

File: test/bundler/bundler_plugin.test.ts [hunk 1]
```diff
@@ -43,6 +42,20 @@ describe("bundler", () => {
       stdout: "foo",
     },
   });
+  itBundled("plugin/ResolveAsync", {
+    files: resolveFixture,
+    plugins(builder) {
+      builder.onResolve({ filter: /\.magic$/ }, async args => {
+        await new Promise(resolve => setTimeout(resolve, 10));
```
Comment (Jarred-Sumner) [inline test/bundler/bundler_plugin.test.ts]:
```suggestion
        await Bun.sleep(10);
```

File: test/js/bun/plugin/plugins.test.ts [hunk 2]
```diff
@@ -95,6 +95,39 @@ plugin({
...
+      loader: "js",
+    }));
+  },
+});
+
+plugin({
+  name: "async resolve",
+  setup(builder) {
+    builder.onResolve({ filter: /original/, namespace: "async-resolve" }, async () => {
+      await new Promise(resolve => setTimeout(resolve, 10));
```
Comment (Jarred-Sumner) [inline test/js/bun/plugin/plugins.test.ts]:
```suggestion
      await Bun.sleep(10);
```

**Comment by Jarred-Sumner** [comment]
Thank you for this

It's technically possible to do this -- but permitting it is a massive performance footgun. Sharp should not be executed during module resolution. It should only be executed onLoad, which fully supports async/await. Multiple files can be loaded simultaneously. Module resolution is (unfortunately) synchronous. If the API allows this, people will get confused (as shown in this PR) 

I'm not confident this makes sense for Bun to support.

**Comment by gungun974** [comment]
> Thank you for this
> 
> It's technically possible to do this -- but permitting it is a massive performance footgun. Sharp should not be executed during module resolution. It should only be executed onLoad, which fully supports async/await. Multiple files can be loaded simultaneously. Module resolution is (unfortunately) synchronous. If the API allows this, people will get confused (as shown in this PR)
> 
> I'm not confident this makes sense for Bun to support.

I was unaware of this difference between the `onResolve` and `onBuild` APIs of plugins. 
The initial reason I went down this implementation route was that I found it strange that Bun could bundle an asynchronous `onResolve` and that `esbuild` could do the same for its `onResolve.` 
However, I can understand if `onResolve` is 100% synchronous and blocking, that this could pose a problem.

Afterwards, I had three reasons for thinking that this PR could be good.

The first reason is that technically what my plugin aims to do is not to load a file with its content, but to indicate to Bun that the import should be done with another file. That's why `onResolve` seemed more coherent to me.

The second reason is the infamous issue with async functions. Once you find yourself with an async function, you can't make it synchronous even if you want to. There's no magical AsyncToSync function.

The third reason is that `onLoad` simply does not support passing a path to the pre-existing `file"` loader. This results in Bun never copying the transformed assets into the build folder when bundling the project, which is very inconvenient in my case.

In any case, here are the initial three reasons why I went with this implementation. 
If this is behavior that Bun should never have at runtime, I can understand. 
On the other hand, that would mean that the phrase `onResolve() doesn't support pending promises yet` should be changed, and that differences could exist between Build and Runtime modes, which in my opinion could confuse other people.

 thank you for looking at my PR @Jarred-Sumner

**Comment by gungun974** [comment]
Of course, just one last thing regarding the use of Sharp in `onResolve`. Obviously, it's a heavy operation, which is why in my plugin I don't regenerate an image that has already been generated.

It's my responsibility (the plugin maker), not really Bun's, to ensure that I'm not running a 15-second process between each Import. (I'm exaggerating with the time, but you get the point.)

**Comment by gungun974** [comment]
# The Discovery
So here some news about the development of my Plugin.

Since you mentioned performance issues, I looked into whether it would be possible to add the missing file loader to Bun from the start. 

So I was working in the source code and starting to write unit tests. While writing these tests, I wondered if I could create an `onLoad` that returns an import function that would import as Bun already does with file. 

It turns out this is possible and actually ***mimics the behavior of `onResolve`, but async,*** with a slightly longer process.

So in this branch of my plugin I made some experiment between the version that used this patch and my workaround with bun 1.0.2 https://github.com/gungun974/bun-image-transform/tree/asyncOnResolveWorkaround

# The benchmarks

For each test, I based myself on this command to run the tests 1000 times and see the performance difference between the PR and directly using `onLoad`, which returns an import.
```bash
bun test --rerun-each 1000
```

## Here the results of Bun 1.0.2 with my work around :
```
 14985 pass
 0 fail
 25974 expect() calls
Ran 14985 tests across 1000 files. [8.18s]
```

## Here the result of Bun 1.0.3 with my patch and the async thing :
```
[...]
tests/plugin.test.ts: (run #108)
[...]
zsh: IOT instruction (core dumped)  ~/bun test --rerun-each 1000
```
If I rerun :
```
[...]
tests/plugin.test.ts: (run #30)
[...]
zsh: segmentation fault (core dumped) 
```
Again :
```
[...]
tests/plugin.test.ts: (run #78)
[...]
GC scan found corrupt object: structureID is zero!
GC type: Eden
Object at: 0x7fa5d1eb5260
Structure ID: 0 (0x7f8e00000000)
Object contents: 0x0000000000000000 0x0000000000000000
Is marked: true
Is newly allocated: true
Block: 0x7fa5d1eb4000
Live:YES, Empty:no, Allocated:no, CanAllocateButNotEmpty:no, Destructible:no, Eden:YES, Unswept:no, MarkingNotEmpty:YES, MarkingRetired:no
Is marked raw: true
Marking version: 2
Heap marking version: 2
Is newly allocated raw: true
Newly allocated version: 128
Heap newly allocated version: 128
zsh: IOT instruction (core dumped) 
```
# So what ?

As you can see, the results for my patch are not great. While normally no one should ever do this with my plugin, the fact that one could theoretically crash Bun is simply not acceptable.

Even though I personally find it odd that we only have async `onResolve` in Bun.build, adding async in runtime with my patch would be far too unstable.

I could try to find ways to speed this up, but as you've seen, it's not worth it. In reality, no one would ever need an async `onResolve`.

However, the reason I'm here today is because in ticket #5314, the error message was `onResolve() doesn't support pending promises yet`. 
This could lead others, like me, to be confused in the future, thinking that Bun plans to implement this feature one day, when in fact that will never be the case.

Also, the reason I did all this was because nowhere in the documentation was it written that it was possible to simply use the following method (see below) to change the loaded file.
```ts
build.onLoad({ filter: /&bunimg$/ }, async (args) => {
    const newPath = await someAwaitThing(args.path);
    
    return {
        contents: `
            export {default} from ${JSON.stringify(newPath)}
        `,
        loader: "js",
    };
});
```
This forces one to think that if they want to load a file, they must go through `onResolve`, and if they can't handle asynchronous behavior, then it's simply impossible. However, as I was able to show you, that's not the case.

# What to do so ?

I propose two modifications that go against the original direction of this PR:

- Change the error message for asynchronous `onResolve` to `onResolve() with pending promises is unstable in Bun Runtime.` this way, no one else would look to modify it **like I did**.

- Secondly, make changes in the `runtime/plugins` section of the documentation so that:
  - It's clearly understood that for performance reasons, `onResolve` cannot be asynchronous, unlike `onLoad`.
  - It's reminded to plugin creators that it's possible to mimic the behavior of asynchronous `onResolve` and the file loader, with a simple JS rexport as I showed you earlier.

These changes could prevent future confusion and guide developers more effectively.

In any case, I'm grateful for your comment and the magic of the 12-hour time difference, which allowed me to conduct this in-depth study of the issue and propose an alternative that will likely suit both you and me, and also benefit the Bun community.

The only question to ask you is whether I should close this PR and open another PR with these fixes, or can I save time and just change all my commits and the title? This would avoid having to recontextualize the whole history.

In short, @Jarred-Sumner, I know this was long to write, but it's just that I want to give my best for this young and incredible community that is Bun. You are doing truly amazing work, and that's why, even if my contribution will be smaller, we will all move forward together!

I await your opinion and whether I should create another PR now. ^^
